### ISSUE TITLE
I trained a model, and it got worse with more images

### ISSUE THREAD

**Issue by DataJuggler** [dotnet/machinelearning-modelbuilder#2064]
I am using Model Builder in VS 2022, CPU because I can't get the GPU version to work, Image Classification project.

I created an image library of 100 images per fonts, with random verbs. It was actually pretty good at recommending with 100 images.
Not always right, but enough I was convinced my prototype would work.

I ran again with 1,000 images per font. Results got worse.

I tried again with 5,000 images per font, and the result are now so bad, it only recommends the first 3, and they all start with A.

So for 1.1 million images, 241 fonts * 5101 images (5,000 phrases per font, 100 large numbers plus 1 full alphabet per font) and then every guess from the model starts with A, and always the 1st 3 A's, after testing about 15 images.

It seems your model doesn't improve with more data, it just starts alphabetically (or first 3, since my folders are alphabetical) and doesn't go much further. With less images it did.

So I will be writing my own Font Detection library it looks like.

![image](https://user-images.githubusercontent.com/26331133/158033207-db11afd4-045f-493d-84ad-3877ef8cb0b1.png)

Maybe someone can tell me why it only takes the first 3 choices, and shows those in different order. The source folder structure looked like this: With 100 images and even 1,000 per font, your guesses were not weighted to the 1st 3 only.

![image](https://user-images.githubusercontent.com/26331133/158033285-56d9b762-e02f-423d-8f89-09c02fc6e148.png)

Just seems your algorithms don't scale well, or maybe 200+ choices x 5,000 images is more than it is designed for, but it worked with less images. Seems opposite of what I expect results to do.

If it doesn't work, what do you do? Just scrap the project?

**Comment by beccamc** [comment]
@LittleLittleCloud Can you comment on this?

**Comment by LittleLittleCloud** [comment]
@DataJuggler 

There's no guarantee that a larger dataset can always produce a better model, For example, a larger dataset might have too much noise, which result in underfitting. In you case, the noise might be little difference among some of fonts.

BTW What's the accuracy of the model trained with the largest dataset. Is that still 100% or much lower

> I tried again with 5,000 images per font, and the result are now so bad, it only recommends the first 3, and they all start with A.

It could also be caching, Can you clear the cache folder (%temp%AutoML-NNI) and launching the training again.

**Comment by DataJuggler** [comment]
The largest dataset has an accuracy of 0. The best results were with 100 images per font. I would estimate 30 - 40 % of the time you at least listed the font in your top 5. With more items, you only recommend the first 3 fonts, so completely useless. 

Thinking about the problem further, I am attempting to divide the image up into characters. For some fonts this is relatively simple due to each create is in it's own rectangle. Other fonts overlap rectangles so those type I put off till version 2 if I get this working. Perhaps if I try to determine the font of a known character it might be get better results. Compare only A's for example.

I don't know if the ML.NET team would wants this, but the large dataset is about a gig. I can put it on my site or somewhere for ML.NET to download if they would like for to improve ML.NET. Seems like a more useful use case than determining if a dog is in a picture or not (the types of examples I have seen).

It is basically 230+ folders with font images that look like:
![image](https://user-images.githubusercontent.com/26331133/158491332-dc384a51-574d-4678-9c4a-b3092ac63d20.png)

I can generate a model of any size, it just takes longer for the 5K per font or larger.

I am 99.99% sure it is not caching issue, because I have an open source project called 'The Wolf' (from Pulp Fiction); I empty my temp folder about 5 times per day.  

The 3D software I use and others fill the Temp folder with so much crap. This is since this morning.

![image](https://user-images.githubusercontent.com/26331133/158492044-25b8a3d7-3f31-410a-b843-72403f278013.png)

I will try your suggestion though, just in case I didn't clear between model training.

Is there any way to tell "What it tries" or how it guesses, and tell it what you are trying is not working?

Thanks for the response

**Comment by LittleLittleCloud** [comment]
The algo model builder used is a pre-trained ResNet with softmax/ log loss as classifier. During training, model builder won't adjust weight in that pre-trained ResNet though, it just trains classifier layer. 

 > Is there any way to tell "What it tries" or how it guesses

The way how the trained model "guesses" is it "guesses" the probability of each catagory (which sum up to 1) and return the label with highest probability. Since your latest dataset has ~250 catagories and what the model does is no better than random guessing ( so probably 1/250 for each catagory), that's probably why you always see the first three is returned. (because each catagory has approximately the same score, the model just return the first three).

> tell it what you are trying is not working?

That's actually happening during training. While training, the model will just "guess" random probabilities for each catagory first and then calculate the distance between the actual label and "guess result", which will be used to adjust the weight in classifier layer. Since the accuracy for training model is still 0, I wonder even if you can "tell" model the actual label, it still won't work.

## What you can do to improve model
In image classification, even the state-of-the-art model can merely beaten human in very specific cases so one thumb of rule is check if YOU as a human can classify your dataset correctly. If not, perhaps the first thing you need to do is to improve your dataset, including but not limited to increase difference among labels, reduce noise ...

**Comment by DataJuggler** [comment]
I even tried a model with Just A or B, and it is not even good at picking out B’s vs A’s in 200+ fonts.

Thank you for the detailed explanation though. I am going to write my own Font Forensics library suited for just this use case.

**Comment by beccamc** [comment]
Thanks for reporting this problem. Looks like the issue has been resolved. If you have any other questions or comments please open a new issue. Thank you!
